server/src/main/java/org/opensearch/rest/pagination/ShardPaginationStrategy.java

```diff
@@ -0,0 +1,244 @@
+        List<IndexMetadata> metadataSublist = tuple.v2();
+        // Get list of index names from the trimmed metadataSublist
+        this.requestedIndices = metadataSublist.stream().map(metadata -> metadata.getIndex().getName()).collect(Collectors.toList());
+        this.pageToken = getResponseToken(
+            pageParams.getSize(),
+            sortedIndices.size(),
+            metadataSublist.isEmpty() ? null : metadataSublist.get(metadataSublist.size() - 1),
+            tuple.v1().isEmpty() ? null : tuple.v1().get(tuple.v1().size() - 1)
+        );
+    }
+
+    private static Predicate<IndexMetadata> getMetadataFilter(String requestedTokenStr, String sortOrder) {
+        boolean isAscendingSort = sortOrder.equals(PARAM_ASC_SORT_VALUE);
+        ShardStrategyToken requestedToken = Objects.isNull(requestedTokenStr) || requestedTokenStr.isEmpty()
+            ? null
+            : new ShardStrategyToken(requestedTokenStr);
+        if (Objects.isNull(requestedToken)) {
+            return indexMetadata -> true;
+        }
+        return metadata -> {
+            if (metadata.getIndex().getName().equals(requestedToken.lastIndexName)) {
+                return true;
+            } else if (metadata.getCreationDate() == requestedToken.lastIndexCreationTime) {
+                return isAscendingSort
+                    ? metadata.getIndex().getName().compareTo(requestedToken.lastIndexName) > 0
+                    : metadata.getIndex().getName().compareTo(requestedToken.lastIndexName) < 0;
+            }
+            return isAscendingSort
+                ? metadata.getCreationDate() > requestedToken.lastIndexCreationTime
+                : metadata.getCreationDate() < requestedToken.lastIndexCreationTime;
+        };
+    }
+
+    private Tuple<List<ShardRouting>, List<IndexMetadata>> getPageData(
+        Map<String, IndexRoutingTable> indicesRouting,
+        List<IndexMetadata> sortedIndices,
+        final int pageSize,
+        String requestedTokenStr
+    ) {
+        List<ShardRouting> shardRoutings = new ArrayList<>();
+        List<IndexMetadata> indexMetadataList = new ArrayList<>();
+        ShardStrategyToken requestedToken = Objects.isNull(requestedTokenStr) || requestedTokenStr.isEmpty()
+            ? null
+            : new ShardStrategyToken(requestedTokenStr);
+        int shardCount = 0;
+        for (IndexMetadata indexMetadata : sortedIndices) {
+            boolean indexShardsAdded = false;
+            Map<Integer, IndexShardRoutingTable> indexShardRoutingTable = indicesRouting.get(indexMetadata.getIndex().getName())
+                .getShards();
+            int shardId = Objects.isNull(requestedToken) ? 0
+                : indexMetadata.getIndex().getName().equals(requestedToken.lastIndexName) ? requestedToken.lastShardId + 1
+                : 0;
+            for (; shardId < indexShardRoutingTable.size(); shardId++) {
+                shardCount += indexShardRoutingTable.get(shardId).size();
```

Comment:
yeah, that's a good point. The only thing is it will require handling in the TransportAction as well, so moved the code which is instantiating this strategy inside a try catch block.